retrievalmarket/impl/clientstates/client_states.go

```diff
@@ -28,133 +26,100 @@ func errorFunc(err error) func(*rm.ClientDealState) {
+		return ctx.Event(rm.ClientEventBadPaymentRequested, "too much money requested for bytes sent")
 	}
+
 	// create payment voucher with node (or fail) for (fundsSpent + paymentRequested)
 	// use correct payCh + lane
 	// (node will do subtraction back to paymentRequested... slightly odd behavior but... well anyway)
-	voucher, err := environment.Node().CreatePaymentVoucher(ctx, deal.PayCh, big.Add(deal.FundsSpent, deal.PaymentRequested), deal.Lane)
+	voucher, err := environment.Node().CreatePaymentVoucher(ctx.Context(), deal.PaymentInfo.PayCh, big.Add(deal.FundsSpent, deal.PaymentRequested), deal.PaymentInfo.Lane)
 	if err != nil {
-		return errorFunc(xerrors.Errorf("creating payment voucher: %w", err))
+		return ctx.Event(rm.ClientEventCreateVoucherFailed, err)
 	}
 
 	// send payment voucher (or fail)
-	err = environment.DealStream().WriteDealPayment(rm.DealPayment{
+	err = environment.DealStream(deal.ID).WriteDealPayment(rm.DealPayment{
 		ID:             deal.DealProposal.ID,
-		PaymentChannel: deal.PayCh,
+		PaymentChannel: deal.PaymentInfo.PayCh,
 		PaymentVoucher: voucher,
 	})
 	if err != nil {
-		return errorFunc(xerrors.Errorf("writing deal payment: %w", err))
+		return ctx.Event(rm.ClientEventWriteDealPaymentErrored, err)
 	}
 
-	// return modify deal function --
-	// status = DealStatusOngoing
-	// paymentRequested = 0
-	// fundsSpent = fundsSpent + paymentRequested
-	// if paymentRequested / pricePerByte >= currentInterval
-	// currentInterval = currentInterval + proposal.intervalIncrease
-	// bytesPaidFor = bytesPaidFor + (paymentRequested / pricePerByte)
-
-	return func(deal *rm.ClientDealState) {
-		if deal.Status == rm.DealStatusFundsNeededLastPayment {
-			deal.Status = rm.DealStatusCompleted
-		} else {
-			deal.Status = rm.DealStatusOngoing
-		}
-		deal.FundsSpent = big.Add(deal.FundsSpent, deal.PaymentRequested)
-		bytesPaidFor := big.Div(deal.PaymentRequested, deal.PricePerByte).Uint64()
-		if bytesPaidFor >= deal.CurrentInterval {
-			deal.CurrentInterval += deal.DealProposal.PaymentIntervalIncrease
-		}
-		deal.BytesPaidFor += bytesPaidFor
-		deal.PaymentRequested = abi.NewTokenAmount(0)
-	}
+	return ctx.Event(rm.ClientEventPaymentSent)
 }
 
 // ProcessNextResponse reads and processes the next response from the provider
-func ProcessNextResponse(ctx context.Context, environment ClientDealEnvironment, deal rm.ClientDealState) func(*rm.ClientDealState) {
+func ProcessNextResponse(ctx fsm.Context, environment ClientDealEnvironment, deal rm.ClientDealState) error {
 	// Read next response (or fail)
-	response, err := environment.DealStream().ReadDealResponse()
+	response, err := environment.DealStream(deal.ID).ReadDealResponse()
 	if err != nil {
-		return errorFunc(xerrors.Errorf("reading deal response: %w", err))
+		return ctx.Event(rm.ClientEventReadDealResponseErrored, err)
 	}
 
 	// Process Blocks
 	totalProcessed := uint64(0)
-	completed := false
+	completed := deal.Status == rm.DealStatusBlocksComplete
```

Comment:
it will be empty but it's a good point -- we really shouldn't be consuming blocks period once all blocks are consumed -- arguably if anything is in response.Blocks it should error at that point.